src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -984,6 +984,253 @@ private Task WriteSuffix()
+            var percentageIndex = pathToEndSpan.IndexOf(BytePercentage);
+
+            var pathEnd = MinNonZero(spaceIndex, questionMarkIndex, percentageIndex);
+
+            if (spaceIndex == -1 && questionMarkIndex == -1 && percentageIndex == -1)
+            {
+                RejectRequestLine(start, end);
+            }
+            else if (percentageIndex != -1)
+            {
+                needDecode = true;
+
+                pathEnd = MinNonZero(spaceIndex, questionMarkIndex);
+                if (questionMarkIndex == -1 && spaceIndex == -1)
+                {
+                    RejectRequestLine(start, end);
+                }
+            }
+
+            var queryString = "";
+            var queryEnd = pathEnd;
+            if (questionMarkIndex != -1)
+            {
+                if (spaceIndex == -1)
+                {
+                    RejectRequestLine(start, end);
+                }
+
+                queryString = pathToEndSpan.Slice(pathEnd, queryEnd).GetAsciiString();
+            }
+
+            if (pathBegin == pathEnd)
+            {
+                RejectRequestLine(start, end);
+            }
+
+            var versionBegin = queryEnd + 1;
+            var versionToEndSpan = pathToEndSpan.Slice(versionBegin, pathToEndSpan.Length - versionBegin);
+            // var ves = Encoding.ASCII.GetString(versionToEndSpan.ToArray());
+            versionBegin = 0;
+            var versionEnd = versionToEndSpan.IndexOf(ByteCR);
+
+            if (versionEnd == -1)
+            {
+                RejectRequestLine(start, end);
+            }
+
+            if (!versionToEndSpan.GetKnownVersion(out string httpVersion))
+            {
+                httpVersion = new Utf8String(versionToEndSpan.Slice(0, versionEnd)).ToString();
+
+                if (httpVersion == string.Empty)
+                {
+                    RejectRequestLine(start, end);
+                }
+                else
+                {
+                    RejectRequest(RequestRejectionReason.UnrecognizedHTTPVersion, httpVersion);
+                }
+            }
+
+            if (versionToEndSpan[versionToEndSpan.Length - 1] != ByteLF)
+            {
+                RejectRequestLine(start, end);
+            }
+
+            var pathBuffer = pathToEndSpan.Slice(pathBegin, pathEnd);
+            var targetBuffer = pathToEndSpan.Slice(pathBegin, queryEnd);
+
+            // URIs are always encoded/escaped to ASCII https://tools.ietf.org/html/rfc3986#page-11
+            // Multibyte Internationalized Resource Identifiers (IRIs) are first converted to utf8;
+            // then encoded/escaped to ASCII  https://www.ietf.org/rfc/rfc3987.txt "Mapping of IRIs to URIs"
+            string requestUrlPath;
+            string rawTarget;
+            if (needDecode)
+            {
+                // Read raw target before mutating memory.
+                rawTarget = targetBuffer.GetAsciiString() ?? string.Empty;
```

Comment:
Ok, the proper check is still in place: https://github.com/aspnet/KestrelHttpServer/pull/1378/files/391393b79db3c14d8a9f877a3bb719f4cae49fcc#diff-4a391b137e1879654de7fd4febc3db67R1120

Don't do `?? string.Empty`. Assert that `rawTarget != null`.